internal/controllers/mbsc_reconciler.go

```diff
@@ -0,0 +1,200 @@
+		)
+}
+
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=modulebuildsignconfigs,verbs=get;list;watch;update;patch;create
+//+kubebuilder:rbac:groups=kmm.sigs.x-k8s.io,resources=modulebuildsignconfigs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups="core",resources=secrets,verbs=get;list;watch
+//+kubebuilder:rbac:groups="core",resources=configmaps,verbs=get;list;watch
+//+kubebuilder:rbac:groups="core",resources=pods,verbs=create;list;watch;delete
+
+// Reconcile lists all nodes and looks for kernels that match its mappings.
+// For each mapping that matches at least one node in the cluster, it creates a DaemonSet running the container image
+// on the nodes with a compatible kernel.
+func (r *mbscReconciler) Reconcile(ctx context.Context, mbscObj *kmmv1beta1.ModuleBuildSignConfig) (ctrl.Result, error) {
+	res := ctrl.Result{}
+
+	logger := log.FromContext(ctx)
+
+	if mbscObj.GetDeletionTimestamp() != nil {
+		// [TODO] delete build/sign pods
+		return res, nil
+	}
+
+	err := r.reconHelperAPI.updateStatusByBuilds(ctx, mbscObj)
+	if err != nil {
+		return res, fmt.Errorf("failed to update MSBC %s status based on the result of the builds: %v", mbscObj.Name, err)
+	}
+
+	err = r.reconHelperAPI.processImagesSpecs(ctx, mbscObj)
+	if err != nil {
+		return res, fmt.Errorf("failed to process images of MSBC %s: %v", mbscObj.Name, err)
+	}
+
+	logger.Info("run garbage collector for build/sign pods")
+	err = r.reconHelperAPI.garbageCollect(ctx, mbscObj)
+	if err != nil {
+		return res, fmt.Errorf("failed to run garbage collector for MBSC %s: %v", mbscObj.Name, err)
+	}
+
+	return res, nil
+}
+
+//go:generate mockgen -source=mbsc_reconciler.go -package=controllers -destination=mock_mbsc_reconciler.go mbscReconcilerHelperAPI
+
+type mbscReconcilerHelperAPI interface {
+	updateStatusByBuilds(ctx context.Context, mbscObj *kmmv1beta1.ModuleBuildSignConfig) error
+	processImagesSpecs(ctx context.Context, mbscObj *kmmv1beta1.ModuleBuildSignConfig) error
+	garbageCollect(ctx context.Context, mbscObj *kmmv1beta1.ModuleBuildSignConfig) error
+}
+
+type mbscReconcilerHelper struct {
+	client       client.Client
+	buildSignAPI buildsign.Manager
+	mbscAPI      mbsc.MBSC
+}
+
+func newMBSCReconcilerHelper(client client.Client, buildSignAPI buildsign.Manager, mbscAPI mbsc.MBSC) mbscReconcilerHelperAPI {
+	return &mbscReconcilerHelper{
+		client:       client,
+		buildSignAPI: buildSignAPI,
+		mbscAPI:      mbscAPI,
+	}
+}
+
+func (mrh *mbscReconcilerHelper) updateStatusByBuilds(ctx context.Context, mbscObj *kmmv1beta1.ModuleBuildSignConfig) error {
+	errs := make([]error, 0, len(mbscObj.Spec.Images))
+	patchFrom := client.MergeFrom(mbscObj.DeepCopy())
+	for _, imageSpec := range mbscObj.Spec.Images {
+		status, err := mrh.buildSignAPI.GetStatus(ctx, mbscObj.Name, mbscObj.Namespace, imageSpec.ModuleImageSpec.KernelVersion, imageSpec.Action, mbscObj)
+		errs = append(errs, err)
+		if err != nil || status == kmmv1beta1.BuildOrSignStatus("") {
```

Comment:
Although the logic is the same, let's put line 131 inside the `if` as there is no point of appending all `nil` errors by default.

Reply:
i will fix it, but notice that the nil errors will still get appended, since the condition is either an error or empty status, which will probably be more common

Reply:
Then how about we make it a bit more clear, like:
```
if err != nil {
     errs = append(errs, err)
      continue
}

if status == kmmv1beta1.buildOrSignStatus("") {
     continue
}
```

?